src/particles.h

```diff
@@ -20,19 +20,269 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "mapnode.h"
+#include "util/serialize.h"
+#include "util/numeric.h"
 
 // This file defines the particle-related structures that both the server and
 // client need. The ParticleManager and rendering is in client/particles.h
 
+namespace ParticleParamTypes {
+	template<typename T> using BlendFunction = T(float,T,T);
+
+	void serializeParameterValue(std::ostream& os, u8 v);
+	void serializeParameterValue(std::ostream& os, u16 v);
+	void serializeParameterValue(std::ostream& os, u32 v);
+	void serializeParameterValue(std::ostream& os, f32 v);
+	void serializeParameterValue(std::ostream& os, v2f v);
+	void serializeParameterValue(std::ostream& os, v3f v);
+
+	void deSerializeParameterValue(std::istream& is, u8&  r);
+	void deSerializeParameterValue(std::istream& is, u16& r);
+	void deSerializeParameterValue(std::istream& is, u32& r);
+	void deSerializeParameterValue(std::istream& is, f32& r);
+	void deSerializeParameterValue(std::istream& is, v2f& r);
+	void deSerializeParameterValue(std::istream& is, v3f& r);
+
+	struct Parameter {
+		virtual void serialize(std::ostream &os) const = 0;
+		virtual void deSerialize(std::istream &is) = 0;
+	};
+
+	template <typename T, size_t PN>
+	struct ParameterWrapper : public Parameter {
+		using ValType = T;
+		using pickFactors = float[PN];
+
+		T val;
+		using This = ParameterWrapper<T, PN>;
+
+		ParameterWrapper() = default;
+		ParameterWrapper(const This& a) : val(a.val) {};
+		template <typename... Args>
+		ParameterWrapper(Args... args) : val(args...) {};
+
+		void serialize(std::ostream &os) const override
+			{ serializeParameterValue  (os, this->val); }
+		void deSerialize(std::istream &is)     override
+			{ deSerializeParameterValue(is, this->val); }
+
+		operator T() const { return val; }
+		T operator = (T b) { return val = b; }
+	};
+
+	template <typename T> T numericalBlend(float fac, T min, T max)
+		{ return min + ((max - min) * fac); }
+
+	template <typename T>
+	struct NumericParameter : public ParameterWrapper<T,1> {
+		using This = NumericParameter<T>;
+
+		template <typename... Args>
+		NumericParameter(Args... args) : ParameterWrapper<T,1>(args...) {};
+
+		This interpolate(float fac, const This against) const {
+			return This(numericalBlend(fac, (T)*this, (T)against));
+		}
+		static This pick(float* f, const This a, const This b) {
+			return a.interpolate(f[0], b);
+		}
+	};
+
+	v2f vectorBlend(float* f, const v2f& a, const v2f& b);
+	v3f vectorBlend(float* f, const v3f& a, const v3f& b);
+
+	template <typename T, size_t N>
+	struct VectorParameter : public ParameterWrapper<T,N> {
+		using This = VectorParameter<T,N>;
+		template <typename... Args>
+		VectorParameter(Args... args) : ParameterWrapper<T,N>(args...) {};
+
+		This interpolate(float fac, const This& against) const {
+			return against.val.getInterpolated(this->val, fac);
+		}
+		static This pick(float* f, const This& a, const This& b) {
+			return This(vectorBlend(f, a.val, b.val));
+		}
+	};
```

Comment:
Although surely convenient I'm not sure if I'm a fan of creating these thin wrappers around scalar/vector types...

Reply:
alright, i've replaced this with overloaded functions